spin_coater.c

```diff
@@ -0,0 +1,196 @@
+#include "spin_coater.h"
+#include "stdint.h"
+#include "hardware/timer.h"
+
+#define NUMBER_OF_WHOLES 2
+#define MAX_RPM_VALUE 6000
+#define MICROSECONDS_IN_MINUTE 60000000
+
+static uint32_t count_of_measurements;
+static uint32_t last_time;
+static uint32_t current_rpm;
+
+void
+rpm_feedback_handler(unsigned int gpio, uint32_t events)
+{
+  if (count_of_measurements == 0) {
+    last_time = time_us_32();
+    count_of_measurements++;
+    return;
+  }
+
+  if (count_of_measurements < NUMBER_OF_WHOLES) {
+    count_of_measurements++;
+    return;
+  }
+
+  uint32_t current_time_us;
+  uint32_t time_elapsed_us;
+  current_time_us = time_us_32();
+  time_elapsed_us = current_time_us - last_time;
+  current_rpm = (MICROSECONDS_IN_MINUTE / time_elapsed_us);
+  count_of_measurements = 0;
+}
+#if CONFIG_SPIN_COATER_DSHOT
+bool
+set_dshot_safe(furnace_context_t *ctx, unsigned dshot)
+{
+  if (dshot < MIN_THROTTLE_COMMAND)
+    dshot = MIN_THROTTLE_COMMAND;
+  else if( dshot > MAX_THROTTLE_COMMAND)
+    dshot = MAX_THROTTLE_COMMAND;
+
+  ctx->spin_coater.dshot_throttle_val = dshot;
+  DEBUG_printf("\n DSHOT set to: %lu!\n", dshot);
+  dshot_send_command(ctx->spin_coater.dshot_throttle_val);
+  return true;
+}
+#elif CONFIG_SPIN_COATER_PWM
+bool
+set_spin_coater_pwm_safe(furnace_context_t* ctx, unsigned duty)
+{
+  if (duty < PWM_490_FREQ_DUTY_MIN || duty > PWM_490_FREQ_DUTY_MAX)
+    return false;
+
+  ctx->spin_coater.pwm_duty = duty;
+  DEBUG_printf("\n PWM set to: %lu!\n", duty);
+  return pwm_set_freq_duty(ctx->spin_coater.pwm_slice_num, PWM_CHAN_A, PWM_490_FREQ, duty);
+}
+#endif
+
+int64_t
+timer_spin_callback(alarm_id_t id, void* user_data)
+{
+  furnace_context_t *ctx = (furnace_context_t *)user_data;
+
+  ctx->spin_coater.spin_state = SPIN_SMOOTH_STOP_REQUESTED;
+  const char* msg = "spin_stopped\n";
+  tcp_server_send_data(
+    ctx, ctx->tcp.client_pcb, (const uint8_t*)msg, strlen(msg));
+
+  return 0;
+}
+
+#if CONFIG_SPIN_COATER_DSHOT
+int scale_dshot_value_when_speeding(furnace_context_t* ctx, int rpm_left, int bias){
+    return (50.0 / ctx->spin_coater.set_rpm) * rpm_left + bias;
+}
+
+int scale_dshot_value_when_stopping(furnace_context_t* ctx, int rpm_left){
+    return 50.0 - (50.0 / ctx->spin_coater.set_rpm) * rpm_left;
+}
+
+absolute_time_t do_dshot_smooth_transition(furnace_context_t* ctx)
+{
```

Comment:
Formatting: return types on a separate line and braces on next line.

Reply:
fixed